website/static/js/registrationUtils.js

```diff
@@ -0,0 +1,764 @@
+};
+/**
+ * Creates a template context for editor type subtemplates. Looks for the data type
+ * in the extension map, and if found instantiates that type's ViewModel. Otherwise
+ * return the bare data Object
+ *
+ * @param {Object} data: data in current editor template scope
+ * @returns {Object|ViewModel}
+ **/
+RegistrationEditor.prototype.context = function(data) {
+    $.extend(data, {
+        save: this.save.bind(this),
+        readonly: this.readonly
+    });
+
+    if (this.extensions[data.type]) {
+        return new this.extensions[data.type](data);
+    }
+    return data;
+};
+/**
+ * Check that the Draft is valid before registering
+ */
+RegistrationEditor.prototype.check = function() {
+    var self = this;
+
+    var proceed = true;
+    $.each(self.flatQuestions(), function(i, question) {
+        if (question.required) {
+            var valid = question.value.isValid();
+            proceed = proceed && valid;
+        }
+    });
+    if (!proceed) {
+
+        bootbox.dialog({
+            title: 'Registration Not Complete',
+            message: 'There are errors in your registration. Please double check it and submit again.',
+            buttons: {
+                success: {
+                    label: 'Ok',
+                    className: 'btn-success',
+                    callback: function() {
+                        self.showValidation(true);
+                    }
+                }
+            }
+        });
+    } else {
+        window.location = self.draft().urls.register_page;
+    }
+};
+/**
+ * Load the next question into the editor, wrapping around if needed
+ **/
+RegistrationEditor.prototype.nextQuestion = function() {
+    var self = this;
+
+    var currentQuestion = self.currentQuestion();
+
+    var questions = self.flatQuestions();
+    var index = $osf.indexOf(questions, function(q) {
+        return q.id === currentQuestion.id;
+    });
+    if (index + 1 === questions.length) {
+        self.currentQuestion(questions.shift());
+    } else {
+        self.currentQuestion(questions[index + 1]);
+    }
+};
+/**
+ * Load the previous question into the editor, wrapping around if needed
+ **/
+RegistrationEditor.prototype.previousQuestion = function() {
+    var self = this;
+
+    var currentQuestion = self.currentQuestion();
+
+    var questions = self.flatQuestions();
+    var index = $osf.indexOf(questions, function(q) {
+        return q.id === currentQuestion.id;
+    });
+    if (index - 1 < 0) {
+        self.currentQuestion(questions.pop());
+    } else {
+        self.currentQuestion(questions[index - 1]);
+    }
+};
+/**
+ * Select a page, selecting the first question on that page
+ **/
+RegistrationEditor.prototype.selectPage = function(page) {
+    var self = this;
+
+    var firstQuestion = page.questions[Object.keys(page.questions)[0]];
+    self.currentQuestion(firstQuestion);
+};
+/**
+ * Update draft primary key and updated time on server response
+ **/
```

Comment:
Looks like last saved time may now be correctly tracking the server save, rather than when the request was sent. :smile: We'll want to verify this is working.